op-challenger/fault/types/types.go

```diff
@@ -60,7 +60,7 @@ type TraceProvider interface {
 	GetPreimage(ctx context.Context, i uint64) (preimage []byte, proofData []byte, err error)
 
 	// AbsolutePreState is the pre-image value of the trace that transitions to the trace value at index 0
-	AbsolutePreState(ctx context.Context) []byte
+	AbsolutePreState(ctx context.Context) (common.Hash, error)
```

Comment:
Is it the pre-image (as it was previously) or is it the hash? It might be that we should provide the hash, but if so, we need to update the comment on this function.

Reply:
We need to be providing the pre-image of the absolute pre-state here.  It should be the output of cannon's `state.EncodeWitness` function (https://github.com/ethereum-optimism/optimism/blob/ba1c236f94e3254fee4d39a488147346d92ef16d/cannon/mipsevm/state.go#L40).  It's the pre-image provided to the step function and the contract will verify that its hash matches the absolute pre-image _claim_ specified when setting up the dispute game type.

The complication here is that the pre-state pre-image we return here is the binary encoded witness but the `--input` state passed when executing cannon needs to be the JSON file which also includes the full memory state (the memory is accessed via the proof data on-chain since it's too large to post it all).

The question is whether we get Cannon to add the binary witness to the state it generates from `load-elf` or if we just add the ability to parse the state JSON and compute the witness to challenger (essentially pull in that state.json either by just copying it or via a module reference). If we're going to verify the pre-state actually matches we probably need the logic in challenger, otherwise we can verify the binary witness matches the pre-state hash but can't verify the state in the JSON actually matches the witness. It kind of sucks to pull so much of cannon into challenger though and is quite possibly going to cause some problematic dependency cycles.

So maybe the best option is to add a new subcommand to Cannon to convert a JSON state to the encoded witness. Then the challenger can be given the state as JSON to use with `--input` and run this subcommand to convert it to the encoded witness, guaranteeing it matches the JSON content.

Reply:
Took the most minimal-diff approach to this whereby the cannon trace provider reads the state.json file and unmarshals it into a cannon `mipsevm.State` which we can just call `EncodeWitness()` on.